Replicator/tests/ReplicatorCollectionTest.cc

```diff
@@ -0,0 +1,251 @@
+            specs.push_back(spec);
+        });
+        return specs;
+    }
+    
+    C4Collection* getCollection(C4Database* db, CollectionSpec spec, bool mustExist =true) {
+        auto coll = db->getCollection(spec);
+        Assert(!mustExist || coll != nil);
+        return coll;
+    }
+    
+    int addDocs(C4Database* db, CollectionSpec spec, int total) {
+        C4Collection* coll = getCollection(db, spec);
+        string prefix = db == this->db ? "db1" : (db == this->db2 ? "db2" : "newdoc");
+        return ReplicatorLoopbackTest::addDocs(coll, 0ms, total, prefix);
+    }
+    
+private:
+    
+    vector<C4ReplicationCollection> replCollections(vector<CollectionSpec>& specs,
+                                                    C4ReplicatorMode pushMode,
+                                                    C4ReplicatorMode pullMode)
+    {
+        vector<C4ReplicationCollection> colls(specs.size());
+        for (int i = 0; i< specs.size(); i++) {
+            colls[i].collection = specs[i];
+            colls[i].push = pushMode;
+            colls[i].pull = pullMode;
+        }
+        return colls;
+    }
+};
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Use Nonexisting Collections", "[Push][Pull]") {
+    vector<CollectionSpec>specs = {CollectionSpec("dummy1"_sl), CollectionSpec("dummy2"_sl)};
+    _expectedError = {LiteCoreDomain, kC4ErrorNotFound};
+    runPushPullReplication(specs, specs);
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Use Unmatched Collections", "[Push][Pull]") {
+    _expectedError = {LiteCoreDomain, kC4ErrorNotFound};
+    runPushPullReplication({Roses, Lavenders}, {Tulips, Lavenders});
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Use Zero Collections", "[Push][Pull]") {
+    _expectedError = {LiteCoreDomain, kC4ErrorInvalidParameter};
+    runPushPullReplication({}, {});
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Push Default Collection", "[Push]") {
+    _expectedDocumentCount = addDocs(db, Default, 10);
+    runPushReplication({Default}, {Default});
+    validateCollectionCheckpoints(db, db2, 0, "{\"local\":10}");
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Pull Default Collection", "[Pull]") {
+    _expectedDocumentCount = addDocs(db, Default, 10);
+    runPullReplication({Default}, {Default});
+    validateCollectionCheckpoints(db2, db, 0, "{\"remote\":10}");
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Push/Pull Default Collection", "[Push][Pull]") {
+    _expectedDocumentCount = addDocs(db, Default, 10);
+    _expectedDocumentCount += addDocs(db2, Default, 5);
+    runPushPullReplication({Default}, {Default});
+    validateCollectionCheckpoints(db, db2, 0, "{\"local\":15,\"remote\":15}");
+}
+
+TEST_CASE_METHOD(ReplicatorCollectionTest, "Push Single Collection", "[Push]") {
+    _expectedDocumentCount = addDocs(db, Guitars, 10);
+    runPushReplication({Guitars}, {Guitars});
+    validateCollectionCheckpoints(db, db2, 0, "{\"local\":10}");
+}
```

Comment:
These are getting a bit out of hand.  I suggest writing one test with sections rather than separate tests for every variation.  There is a handy `SECTION` keyword.  Alternatively, I know there are other tests that run in different variations for every test in the class, which seems like it would be applicable here.

Reply:
OK